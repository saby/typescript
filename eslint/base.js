'use strict';

const es6Rules = require('./rules/es6');
const tsRules = require('./rules/ts');
const tsxRules = require('./rules/tsx');
const testTsxRules = require('./rules/test-tsx');
const baseJSRules = require('./rules/basejs');
const sbisRules = require('./rules/sbisRules');

// Configuration for any JS code: es5, es6+, node.js
module.exports = {
   parser: '@typescript-eslint/parser',
   parserOptions: {
      ecmaVersion: 13,
      sourceType: 'script'
   },
   plugins: [
      '@typescript-eslint',
   ],
   env: {
      browser: true,
      node: true,
      es6: true,
      amd: true,
      mocha: true,
      jest: true,
      jquery: true
   },
   globals: {
      rk: true
   },
   rules: {

      // ВСЕ правила, что находятся здесь, совпадают и для js, и для ts.
      // Если что-то не совпадает - в отдельный конфиг

      ...sbisRules,

      // Possible Errors
      'no-debugger': 'error',
      'no-empty': 'error',
      'no-unsafe-finally': 'error',
      'use-isnan': 'error',

      // Best Practices
      'dot-notation': ['error', { allowKeywords: true }],
      eqeqeq: ['error', 'always', { null: 'ignore' }],
      'guard-for-in': 'error',
      'no-caller': 'error',
      'no-eval': 'error',
      'no-new-wrappers': 'error',
      'no-return-await': 'error',
      'no-sequences': 'error',
      radix: 'error',

      // Variables
      'no-undef-init': 'error',

      // ECMAScript 6
      //---------------------
      // Settings that encourage us to write code incompatible with ES5 are in ./es6.js
      'constructor-super': 'error'
   },
   overrides: [
      {
         files: ['**/*.js', '**/*.es'],
         rules: baseJSRules
      },
      {

         // этот override должен быть ниже обычного js, чтобы перекрыть его
         files: ['**/*.test.js', '**/*.case.js'],
         rules: {
            'max-nested-callbacks': 'off'
         }
      },
      {
         files: ['**/*.es'],
         parserOptions: {
            ecmaVersion: 9,
            sourceType: 'module'
         },
         rules: es6Rules
      },
      {
         files: ['**/*.ts', '**/*.tsx'],
         rules: tsRules
      },
      {
         files: ['**/*.tsx'],
         rules: tsxRules,
         ecmaFeatures: {
            jsx: true
         },
         plugins: [
            'react',
            'react-hooks'
         ],
         settings: {
            react: {
               pragma: 'React',
               version: '17.0',
            }
         }
      },
      {
         files: ['**/*.test.tsx'],
         rules: testTsxRules,
         ecmaFeatures: {
            jsx: true
         },
         plugins: [
            'testing-library'
         ],
         settings: {
            react: {
               pragma: 'React',
               version: '17.0',
            }
         }
      },
   ]
};
