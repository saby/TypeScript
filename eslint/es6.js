'use strict';

const es6Rules = require('./impl/es6-rules');

// Configuration that forces us to use ES6+
module.exports = {
   extends: './base.js',
   parserOptions: {
      ecmaVersion: 9,
      sourceType: 'module'
   },
   rules: es6Rules
};
