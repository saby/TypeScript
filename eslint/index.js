'use strict';

// Configuration that suitable to any js code: es5, es6+, node.js
module.exports = {
   extends: './base.js',
};
