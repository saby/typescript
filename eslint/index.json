{
   "extends": "./node_modules/saby-typescript/eslint/base.js"
}
