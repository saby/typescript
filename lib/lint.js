// TSlint re-export
module.exports = require('tslint');
