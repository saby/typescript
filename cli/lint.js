#!/usr/bin/env node
// TODO: тут прогон еслинта должен быть в итоге
require('tslint/lib/tslintCli');
