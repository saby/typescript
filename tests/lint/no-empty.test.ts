import { ESLint } from 'eslint';
import { lint } from './utils/lint';
import { getConfig } from './utils/getConfig';

describe('no-empty', () => {
    let linter: ESLint;

    beforeAll(() => {
        linter = new ESLint(getConfig());
    });

    const header = `const foo = 1;
function doSomething(): void {
    // not empty
}
`;

    describe('ошибки', () => {
        const cases = [
            `${header}if (foo) {
}`,
            `${header}while (foo) {
}`,
            `${header}switch(foo) {
}`,
            `${header}try {
    doSomething();
} catch(ex) {

} finally {

}`
        ];

        cases.forEach((text, index) => {
            it(`индекс теста: ${index}`, async () => {
                const result = await lint(linter, text);

                expect(result.messages).toMatchSnapshot();
            });
        });
    });

    describe('не ошибки', () => {
        const cases = [
            `${header}if (foo) {
    // empty
}`,
            `${header}while (foo) {
    /* empty */
}`,
            `${header}try {
    doSomething();
} catch (ex) {
    // continue regardless of error
}`,
            `${header}try {
    doSomething();
} finally {
    /* continue regardless of error */
}`
        ];

        cases.forEach((text, index) => {
            it(`индекс теста: ${index}`, async () => {
                const result = await lint(linter, text);

                expect(result.messages).toMatchSnapshot();
            });
        });
    });
});
