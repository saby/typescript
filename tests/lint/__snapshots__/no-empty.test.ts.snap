// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no-empty не ошибки индекс теста: 0 1`] = `Array []`;

exports[`no-empty не ошибки индекс теста: 1 1`] = `Array []`;

exports[`no-empty не ошибки индекс теста: 2 1`] = `Array []`;

exports[`no-empty не ошибки индекс теста: 3 1`] = `Array []`;

exports[`no-empty ошибки индекс теста: 0 1`] = `
Array [
  Object {
    "column": 10,
    "endColumn": 2,
    "endLine": 6,
    "line": 5,
    "message": "Empty block statement.",
    "messageId": "unexpected",
    "nodeType": "BlockStatement",
    "ruleId": "no-empty",
    "severity": 2,
  },
]
`;

exports[`no-empty ошибки индекс теста: 1 1`] = `
Array [
  Object {
    "column": 13,
    "endColumn": 2,
    "endLine": 6,
    "line": 5,
    "message": "Empty block statement.",
    "messageId": "unexpected",
    "nodeType": "BlockStatement",
    "ruleId": "no-empty",
    "severity": 2,
  },
]
`;

exports[`no-empty ошибки индекс теста: 2 1`] = `
Array [
  Object {
    "column": 1,
    "endColumn": 2,
    "endLine": 6,
    "line": 5,
    "message": "Empty switch statement.",
    "messageId": "unexpected",
    "nodeType": "SwitchStatement",
    "ruleId": "no-empty",
    "severity": 2,
  },
]
`;

exports[`no-empty ошибки индекс теста: 3 1`] = `
Array [
  Object {
    "column": 13,
    "endColumn": 2,
    "endLine": 9,
    "line": 7,
    "message": "Empty block statement.",
    "messageId": "unexpected",
    "nodeType": "BlockStatement",
    "ruleId": "no-empty",
    "severity": 2,
  },
  Object {
    "column": 11,
    "endColumn": 2,
    "endLine": 11,
    "line": 9,
    "message": "Empty block statement.",
    "messageId": "unexpected",
    "nodeType": "BlockStatement",
    "ruleId": "no-empty",
    "severity": 2,
  },
]
`;
